File: objectModel/Python/cdm/objectmodel/cdm_corpus_context.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Any, Dict, List, Optional, TYPE_CHECKING

from cdm.enums import CdmStatusLevel
from cdm.utilities import EventCallback
from cdm.utilities.logging.event_list import EventList

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusDefinition, CdmDocumentDefinition
    from cdm.resolvedmodel import ResolvedAttributeSetBuilder, ResolvedTraitSet


class CdmCorpusContext:
    def __init__(self, corpus: 'CdmCorpusDefinition', status_event: 'EventCallback', \
                    report_at_level: Optional['CdmStatusLevel'] = None, \
                    correlation_id: Optional[str] = None) -> None:
        self.corpus = corpus  # type: CdmCorpusDefinition
        self.report_at_level = report_at_level or CdmStatusLevel.WARNING  # type: CdmStatusLevel
        self.status_event = status_event  # type: EventCallback
        self.events = EventList() # type: EventList
        self.suppressed_log_codes = set() # type: set
        self.feature_flags = {} # type: Dict[str, object]

        self.correlation_id = correlation_id or None # type: Optional[str]

        # --- internal ---
        self._attribute_cache = {}  # type: Dict[str, ResolvedAttributeSetBuilder]
        self._trait_cache = {}  # type: Dict[str, ResolvedTraitSet]
        self._relative_path = None  # type: Optional[str]

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/cdm/CdmCorpusContext.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.cdm;

import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmStatusLevel;
import com.microsoft.commondatamodel.objectmodel.utilities.EventCallback;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.EventList;

import java.util.HashSet;
import java.util.Map;

public interface CdmCorpusContext {

  CdmCorpusDefinition getCorpus();

  void setCorpus(CdmCorpusDefinition value);

  CdmStatusLevel getReportAtLevel();

  void setReportAtLevel(CdmStatusLevel value);

  EventCallback getStatusEvent();

  HashSet<CdmLogCode> getSuppressedLogCodes();

  void setStatusEvent(EventCallback value);

  /**
   * Collects events emitted by the SDK.
   * @return the events
   */
  EventList getEvents();

  /**
   * Returns (optional) correlation ID to be stamped on all recorded status events.
   * @return the correlation ID
   */
  String getCorrelationId();

  /**
   * Sets correlation ID to be stamped on all recorded status events.
   * @param correlationId the correlation ID
   */
  void setCorrelationId(String correlationId);

  /**
   * Set feature flags in context.
   * @param featureFlags map of feature flag name and value.
   */
  void setFeatureFlags(Map<String, Object> featureFlags);

  /**
   * Returns feature flags set in the context.
   * @return Map of feature flag name and value.
   */
  Map<String, Object> getFeatureFlags();

}
